### PR TITLE
Capture main view touch events when drawer is open

### PR DESCRIPTION
**Pull request by realjamie**
In this PR I tried to solve this issue: https://github.com/root-two/react-native-drawer/issues/68

My idea is to let the main-view-container capture the touch-events when the drawer is open. In the original codes, drawer-container and main-view-container shared the same `onStartShouldSetPanResponder`, `onStartShouldSetPanResponderCapture`. This was not correct.

When the drawer is open, the drawer-container should not capture the touch-events, however, the main-view-container SHOULD capture the touch-events in order to block touches from child components on the main-view-container. Thus, the key changes I made in this PR is in `processShouldSet()`:

```
  if (inMask) {
      let toggled = this.processTapGestures()
      if (toggled) return viewType === VIEW_TYPE_DRAWER ? false : true  // The key change I made
      if (this.props.captureGestures && this.props.acceptPan) return true
    }
```

The original logics are quite complicated so I'm not sure whether my changes break anything, though my changes are simple. Would like to ask @rt2zz to verify that! It works well on my drawer:

```
    <Drawer
        acceptPan
        captureGestures
        content={this._renderMenu()}
        negotiatePan={false}
        onClose={this.props.onClose}
        onOpen={this.props.onOpen}
        openDrawerOffset={0.1}
        panCloseMask={1}
        panOpenMask={0}
        panStartCompensation
        ref="drawer"
        styles={stylesRaw.drawer}
        tapToClose
        type="displace"
        tweenDuration={350}
        tweenHandler={this._handleTween}
        tweenEasing={'linear'}>
          {this.props.children}
      </Drawer>
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -301,6 +303,7 @@ class Drawer extends Component {
 
   open = () => {
     this.props.onOpenStart && this.props.onOpenStart()
+    if (this._activeTween) return
```
Comment (realjamie) [inline index.js:306]: I didn't make this change in this PR.

File: index.js [hunk 2]
```diff
@@ -324,6 +327,7 @@ class Drawer extends Component {
 
   close = () => {
     this.props.onCloseStart && this.props.onCloseStart()
+    if (this._activeTween) return
```
Comment (realjamie) [inline index.js:330]: I didn't make this change in this PR.

File: index.js [hunk 3]
```diff
@@ -476,11 +484,12 @@ class Drawer extends Component {
     let viewport = e.nativeEvent.layout
     let oldViewport = this.state.viewport
     if (viewport.width === oldViewport.width && viewport.height === oldViewport.height) return
-    this.resync(viewport)
+    let didRotationChange = viewport.width !== oldViewport.width
+    this.resync(viewport, null, didRotationChange)
```
Comment (realjamie) [inline index.js:488]: I didn't make these changes in this PR.

File: index.js [hunk 4]
```diff
@@ -476,11 +484,12 @@ class Drawer extends Component {
...
   }
 
-  resync(viewport, props) {
-    if (viewport) this._syncAfterUpdate = true
+  resync(viewport, props, didRotationChange) {
+    if (didRotationChange) this._syncAfterUpdate = true
```
Comment (realjamie) [inline index.js:492]: I didn't make these changes in this PR.

**Comment by rt2zz** [comment]
Ok I think I understand the need for this. but I think I might have a alternative solution. What if captureGestures only applied to the main when open and only apply to the drawer when closed. I think this would achieve the same functionality but be a little more generic. I cannot think of a use case where one would want to captureGestures in the other combinations.

**Comment by rt2zz** [comment]
I will take this in and tinker with it a bit...

**Comment by rt2zz** [comment]
@realjamie I took this in, but ended up heavily modifying it. I believe if you set `captureGestures='open'` it should only capture when the drawer is open. Also captureGestures is now much more aggresive in the capture logic and should work better now. Let me know how that goes! (`npm i react-native-drawer@next`)

**Comment by realjamie** [comment]
Hi @rt2zz , thanks for updating! I tried `npm i react-native-drawer@next` but got an error like below. Do you know what happened?
simulator screen shot mar 5 2016 10 02 42 pm

**Comment by rt2zz** [comment]
:tired_face: that should not have been a static class property. I fixed that, removed some console.log and improved the landscape orientation handling, and worked on captureGestures a bit more. new version out on @next tag.

**Comment by realjamie** [comment]
I tried the new version. When I click the button (on the main view) which is supposed to open the drawer, the drawer opens and then immediately closes. The drawer should open only.

**Comment by rt2zz** [comment]
Odd, I tried your Drawer code from above and it is working in my application. There may be something weird going on within your onClose handler. A lot of things changed so it is hard to definitively say what change is causing this. 

Can you share your onClose/onOpen functions?

**Comment by realjamie** [comment]
Will do later today!

**Comment by realjamie** [comment]
Hi, I didn't define onClose / onOpen functions. The button I used to open my drawer calls `this.refs.menu.toggle()`, instead of `this.refs.menu.open()`. Is that the reason?

**Comment by realjamie** [comment]
Oh... I indeed defined onOpen handler, but I thought it won't have effects.

```
  _handleMenuOpen() {
    this.setState({ pressedMenuItem: null });
  }
```

I used pressedMenuItem to display which item I clicked on my drawer. If I removed this onOpen handler, the drawer works well. But I don't understand why this matters..?

**Comment by rt2zz** [comment]
hm, the setState will cause a re-render, which would indicate to me that there is some bug in the drawer related to re-renders. I will continue to dig deeper this week and report back if I figure out what is going on.

**Comment by realjamie** [comment]
Thanks :-)

**Comment by rt2zz** [comment]
@realjamie I added you as a collaborator. No pressure, just thought that might make getting your PR's in a little easier :+1:

**Comment by realjamie** [comment]
got it!

**Comment by realjamie** [comment]
By the way, @rt2zz , how does this fix go?

**Comment by rt2zz** [comment]
haha sorry I forgot to update. I was able to recreate the problem, which iirc was caused by some bad logic in the declarative mode implementation. This I believe was the specific fix: https://github.com/root-two/react-native-drawer/commit/67431e7159eaddf0ae6ea3fedfb2805174861681#diff-168726dbe96b3ce427e7fedce31bb0bcR121

**Comment by realjamie** [comment]
Thanks for updates!

**Comment by rt2zz** [comment]
overall I think v2 is very close. Waiting to get a little more feedback before flipping the switch on npm though.

**Comment by realjamie** [comment]
By the way, I'm not able to do `npm install react-native-drawer@next`, got `invalid`

**Comment by realjamie** [comment]
I pointed it to master and now it works

**Comment by hector** [comment]
Hi, I know this is is marked as closed but could anyone write an example of how to achieve the described behaviour? I cannot get it working.
Thanks in advance.

**Comment by rt2zz** [comment]
👍 there may not be an active release on the next tag anymore. Ill try to get a new major release out this week with Animated

**Pull request merged**